src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,366 @@
+          (col === row || col + row === kWidth - 1)
+        ) {
+          texelValueBytes = kValidPixelColor;
+        } else if (renderShape === RenderShape.BottomLeftTriangle && row > col) {
+          texelValueBytes = kValidPixelColor;
+        } else if (
+          renderShape === RenderShape.ConcaveShape &&
+          (row > col || row + col >= kWidth - 1)
+        ) {
+          texelValueBytes = kValidPixelColor;
+        } else if (renderShape === RenderShape.Square) {
+          texelValueBytes = kValidPixelColor;
+        } else {
+          texelValueBytes = kInvalidPixelColor;
+        }
+
+        const byteOffset = row * bytesPerRow + col * texelValueBytes.byteLength;
+        arrayBuffer.set(texelValueBytes, byteOffset);
+      }
+    }
+    return arrayBuffer;
+  }
+}
+
+export const g = makeTestGroup(IndexFormatTest);
+
+g.test('index_format,uint16')
+  .desc('Test rendering result of indexed draw with index format of uint16.')
+  .params([
+    { indexOffset: 0, _expectedShape: RenderShape.Square },
+    { indexOffset: 6, _expectedShape: RenderShape.BottomLeftTriangle },
+    { indexOffset: 18, _expectedShape: RenderShape.Nothing },
+  ])
+  .fn(t => {
+    const { indexOffset, _expectedShape } = t.params;
+
+    // If this is interpreted as uint32, it will have index 1 and 2 be both 0 and render nothing.
+    // And the index buffer size - offset must be not less than the size required by triangle
+    // list, otherwise it also render nothing.
+    const indexArray = new Uint16Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint16');
+    const result = t.run(indexBuffer, indexArray.length, 'uint16', indexOffset);
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(_expectedShape);
+    t.expectContents(result, expectedTextureValues);
+  });
+
+g.test('index_format,uint32')
+  .desc('Test rendering result of indexed draw with index format of uint32.')
+  .params([
+    { indexOffset: 0, _expectedShape: RenderShape.Square },
+    { indexOffset: 12, _expectedShape: RenderShape.BottomLeftTriangle },
+    { indexOffset: 36, _expectedShape: RenderShape.Nothing },
+  ])
+  .fn(t => {
+    const { indexOffset, _expectedShape } = t.params;
+
+    // If this is interpreted as uint16, then it would be 0, 1, 0, ... and would draw nothing.
+    // And the index buffer size - offset must be not less than the size required by triangle
+    // list, otherwise it also render nothing.
+    const indexArray = new Uint32Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint32');
+    const result = t.run(indexBuffer, indexArray.length, 'uint32', indexOffset);
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(_expectedShape);
+    t.expectContents(result, expectedTextureValues);
+  });
+
+g.test('primitive_restart')
+  .desc('Test primitive restart with each primitive topology.')
```

Comment:
Add a TODO to try to test indexing into a vertex buffer that actually has `2**32` vertices.

Reply:
See comment above for a way to do this without a vertex buffer.